lib/internal/quic/quic.js

```diff
@@ -0,0 +1,2493 @@
+    if (!this.state.isStreamOpenAllowed) {
+      throw new ERR_QUIC_OPEN_STREAM_FAILED();
+    }
+    const handle = this.#handle.openStream(STREAM_DIRECTION_BIDIRECTIONAL);
+    if (handle === undefined) {
+      throw new ERR_QUIC_OPEN_STREAM_FAILED();
+    }
+    const stream = new Stream(this.#streamConfig, handle, this, 0);
+    ArrayPrototypePush(this.#streams, stream);
+    return stream;
+  }
+
+  /**
+   * @returns {Stream}
+   */
+  openUnidirectionalStream() {
+    if (this.#isClosedOrClosing) {
+      throw new ERR_INVALID_STATE('Session is closed');
+    }
+    if (!this.state.isStreamOpenAllowed) {
+      throw new ERR_QUIC_OPEN_STREAM_FAILED();
+    }
+    const handle = this.#handle.openStream(STREAM_DIRECTION_UNIDIRECTIONAL);
+    if (handle === undefined) {
+      throw new ERR_QUIC_OPEN_STREAM_FAILED();
+    }
+    const stream = new Stream(this.#streamConfig, handle, this, 1);
+    ArrayPrototypePush(this.#streams, stream);
+    return stream;
+  }
+
+  /**
+   * Initiate a key update.
+   */
+  updateKey() {
+    if (this.#isClosedOrClosing) {
+      throw new ERR_INVALID_STATE('Session is closed');
+    }
+    this.#handle.updateKey();
+  }
+
+  /**
+   * Gracefully closes the session. Any streams created on the session will be
+   * allowed to complete gracefully and any datagrams that have already been
+   * queued for sending will be allowed to complete. Once all streams have been
+   * completed and all datagrams have been sent, the session will be closed.
+   * New streams will not be allowed to be created. The returned promise will
+   * be resolved when the session closes, or will be rejected if the session
+   * closes abruptly due to an error.
+   * @returns {Promise<void>}
+   */
+  close() {
+    if (!this.#isClosedOrClosing) {
+      this.#isPendingClose = true;
+      this.#handle?.gracefulClose();
+    }
+    return this.closed;
+  }
+
+  /**
+   * A promise that is resolved when the session is closed, or is rejected if
+   * the session is closed abruptly due to an error.
+   * @type {Promise<void>}
+   */
+  get closed() { return this.#pendingClose.promise; }
+
+  /** @type {boolean} */
+  get destroyed() { return this.#handle === undefined; }
+
+  /**
+   * Forcefully closes the session abruptly without waiting for streams to be
+   * completed naturally. Any streams that are still open will be immediately
+   * destroyed and any queued datagrams will be dropped. If an error is given,
+   * the closed promise will be rejected with that error. If no error is given,
+   * the closed promise will be resolved.
+   * @param {any} error
+   */
+  destroy(error) {
+    // TODO(@jasnell): Implement.
+  }
+
+  /**
+   * Send a datagram. The id of the sent datagram will be returned. The status
+   * of the sent datagram will be reported via the datagram-status event if
+   * possible.
+   *
+   * If a string is given it will be encoded as UTF-8.
+   *
+   * If an ArrayBufferView is given, the view will be copied.
```

Comment:
The copy here would be problematic for perf reasons. Is this necessary?

Reply:
This will likely be changed to detach the underlying ArrayBuffer instead of copy. The issue is that we do not want the buffer to be mutated while it is pending to be sent.

Reply:
Maybe add a TODO so we don't forget?